qt/RpcQueue.cc

```diff
@@ -0,0 +1,113 @@
+/*
+ * This file Copyright (C) 2015 Mnemosyne LLC
+ *
+ * It may be used under the GNU GPL versions 2 or 3
+ * or any future license endorsed by Mnemosyne LLC.
+ *
+ * $Id$
+ */
+
+#include "RpcQueue.h"
+
+#include <signal.h>
+
+namespace
+{
+  /*
+   * A helper function to get the future's result if it exists or construct
+   * a dummy one if the previous step's result is an empty future. The rules
+   * of actually having the result and not returning RpcResponse() except for
+   * the first step are checked in stepFinished()
+   */
+  RpcResponse response (QFuture<RpcResponse> f)
+  {
+    if (f.isResultReadyAt (0))
+      return f.result ();
+    else
+      return RpcResponse ();
+  }
+}
+
+RpcQueue::RpcQueue (QObject * parent)
+  : QObject (parent)
+{
+  connect (&myFutureWatcher, SIGNAL (finished ()), SLOT (stepFinished ()));
+}
+
+void
+RpcQueue::add (RpcQueue::QueuedFunction func, RpcQueue::ErrorHandlerFunction onError)
+{
+  myQueue.enqueue (std::make_pair ([func] (QFuture<RpcResponse> f) { return func (response (f)); },
+                                   onError
+                                   ? [onError] (QFuture<RpcResponse> f) { onError (response (f)); }
+                                   : InternalErrorHandlerFunction ()));
+}
+
+
+void
+RpcQueue::add_aux (RpcQueue::QueuedFunctionNoResult func)
+{
+  myQueue.enqueue (std::make_pair ([func] (QFuture<RpcResponse> f) { func (response (f)); return f; },
+                                   InternalErrorHandlerFunction ()));
+}
+
+QFuture<RpcResponse>
+RpcQueue::future ()
+{
+  return myPromise.future ();
+}
+
+void
+RpcQueue::stepFinished ()
+{
+  RpcResponse result;
+
+  if (myFutureWatcher.future ().isResultReadyAt (0))
+    {
+      result = myFutureWatcher.result ();
+      QFuture<RpcResponse> future = myFutureWatcher.future ();
+
+      if (!result.success && result.networkError != QNetworkReply::NoError)
+        {
+          // we can't handle network errors, abort queue and pass the error upwards
+          myPromise.reportFinished (&result);
+          deleteLater ();
```

Comment:
Looks like we'd expect the queue to still be alive if `myTolerateErrors` is true few lines below. Is there a `return` missing?